docs/modularization.md

```diff
@@ -0,0 +1,168 @@
+    (...)
+---
+name: mylib_enum1
+typedef: 'enum'
+    (...)
+```
+
+Each document in the YAML file describes one extern type.
+
+You can also check these files for some real working examples of what you can do with it:
+
+- [`date.yaml`](https://github.com/cross-language-cpp/djinni-support-lib/blob/main/test-suite/djinni/vendor/third-party/date.yaml)
+- [`duration.yaml`](https://github.com/cross-language-cpp/djinni-support-lib/blob/main/test-suite/djinni/vendor/third-party/duration.yaml)
+
+## Usage
+
+To use a library type in your project simply include it in your IDL file and refer to it using its name identifier:
+
+```
+@extern "mylib.yaml"
+
+client_interface = interface +c {
+  foo(): mylib_record1;
+}
+```
+
+## Defining custom types
+
+The YAML files can be created by hand as long as you follow the required format.
+This allows you to support types not generated by Djinni.
+
+See [`duration.yaml`](https://github.com/cross-language-cpp/djinni-support-lib/blob/main/test-suite/djinni/vendor/third-party/duration.yaml)
+and the accompanying translators in [`Duration-objc.hpp`](https://github.com/cross-language-cpp/djinni-support-lib/blob/main/test-suite/handwritten-src/cpp/Duration-objc.hpp),
+[`Duration-jni.hpp`](https://github.com/cross-language-cpp/djinni-support-lib/blob/main/test-suite/handwritten-src/cpp/Duration-jni.hpp)
+and [`Duration-cs.hpp`](https://github.com/cross-language-cpp/djinni-support-lib/blob/main/test-suite/handwritten-src/cpp/Duration-cs.hpp) for an advanced example.
+
+Handwritten translators implement the following concept:
+
+```cpp
+// For C++ <-> Objective-C
+struct Record1
+{
+    using CppType = ::mylib::Record1;
+    using ObjcType = MLBRecord1*;
+
+    static CppType toCpp(ObjcType o) { return /* your magic here */; }
+    static ObjcType fromCpp(CppType c) { return /* your magic here */; }
+
+    // Option 1: use this if no boxing is required
+    using Boxed = Record1;
+    // Option 2: or this if you do need dedicated boxing behavior
+    struct Boxed
+    {
+        using ObjcType = MLBRecord1Special*;
+        static CppType toCpp(ObjcType o) { return /* your magic here */; }
+        static ObjcType fromCpp(CppType c) { return /* your magic here */; }
+    }
+};
+```
+
+```cpp
+// For C++ <-> JNI
+#include "djinni_support.hpp"
+struct Record1
+{
+    using CppType = ::mylib::Record1;
+    using JniType = jobject;
+
+    static CppType toCpp(JniType j) { return /* your magic here */; }
+    // The return type *must* be LocalRef<T> if T is not a primitive!
+    static ::djinni::LocalRef<jobject> JniType fromCpp(CppType c) { return /* your magic here */; }
+
+    using Boxed = Record1;
+};
+```
+
+```cpp
+// For C++ <-> C++/CLI
+struct Record1 {
+    using CppType = ::mylib::Record1;
+    using CsType = ::mylib::cs::Record1;
+
+    static CppType ToCpp(CsType dt) { return /* your magic here */; }
+    static CsType FromCpp(CppType dt) { return /* your magic here */; }
+};
+```
```

Comment:
I sent you a PR (https://github.com/jothepro/djinni-generator-1/pull/1) which is not far from what you added here. However, I do have questions regarding the examples which I explained in that PR. Feel free to bring that content here if you'd like to keep the discussions in a single place.